### PR TITLE
LibJS+LibWeb: Implement various missing things to get a Vue.js hello world running

### PR DESCRIPTION
**Pull request by linusg**
Not quite there yet, it currently chokes on missing `cloneNode()` - but this is getting crowded already, so here's a first PR.

### CODE REVIEW

File: Userland/Libraries/LibWeb/DOM/Document.cpp [hunk 1]
```diff
@@ -628,6 +630,60 @@ NonnullRefPtr<Range> Document::create_range()
...
+        event = Event::create(""); // FIXME: Create FocusEvent
+    } else if (interface.matches("hashchangeevent")) {
+        event = Event::create(""); // FIXME: Create HashChangeEvent
+    } else if (interface.matches("htmlevents")) {
+        event = Event::create("");
+    } else if (interface.matches("keyboardevent")) {
+        event = Event::create(""); // FIXME: Create KeyboardEvent
+    } else if (interface.matches("messageevent")) {
+        event = Event::create(""); // FIXME: Create MessageEvent
+    } else if (interface.matches("mouseevent") || interface.matches("mouseevents")) {
```
Comment (Lubrsi) [inline Userland/Libraries/LibWeb/DOM/Document.cpp:661]: As a shorthand for this, we can do:
```suggestion
    } else if (interface.is_one_of("mouseevent", "mouseevents")) {
```

File: Userland/Libraries/LibWeb/DOM/Document.cpp [hunk 2]
```diff
@@ -628,6 +630,60 @@ NonnullRefPtr<Range> Document::create_range()
...
+        event = UIEvents::MouseEvent::create("", 0, 0);
+    } else if (interface.matches("storageevent")) {
+        event = Event::create(""); // FIXME: Create StorageEvent
+    } else if (interface.matches("svgevents")) {
+        event = Event::create("");
+    } else if (interface.matches("textevent")) {
+        event = Event::create(""); // FIXME: Create CompositionEvent
+    } else if (interface.matches("touchevent")) {
+        event = Event::create(""); // FIXME: Create TouchEvent
+    } else if (interface.matches("uievent") || interface.matches("uievents")) {
```
Comment (Lubrsi) [inline Userland/Libraries/LibWeb/DOM/Document.cpp:671]: Ditto

File: Userland/Libraries/LibWeb/DOM/Document.cpp [hunk 3]
```diff
@@ -628,6 +630,60 @@ NonnullRefPtr<Range> Document::create_range()
     return Range::create(*this);
 }
 
+// https://dom.spec.whatwg.org/#dom-document-createevent
+NonnullRefPtr<Event> Document::create_event(const String& interface)
```
Comment (Lubrsi) [inline Userland/Libraries/LibWeb/DOM/Document.cpp:634]: `interface` should be made all lowercase, as it needs to be a case-insensitive match.

Comment (linusg) [reply Userland/Libraries/LibWeb/DOM/Document.cpp:634]: `matches()` defaults to `CaseSensitivity::CaseInsensitive`

Comment (Lubrsi) [reply Userland/Libraries/LibWeb/DOM/Document.cpp:634]: Oh, didn't know that :)
I just realised is_one_of is case sensitive so it's probably better to leave it all as matches then.

Comment (linusg) [reply Userland/Libraries/LibWeb/DOM/Document.cpp:634]: now that I think about it, matches is probably the wrong method to use here as it's really intended for globs. `equals_ignoring_case()` would do the trick but just uses `to_lowercase` under the hood, so we'd call that a bunch - so we're back to your initial suggestion :smile:

**Comment by Lubrsi** [comment]
With this, web platform tests can now finish setting up and gives some output:
Screenshot from 2021-04-10 18-25-53

File: Userland/Libraries/LibWeb/DOM/Document.cpp (hunk 3 above)
Comment (linusg) [reply Userland/Libraries/LibWeb/DOM/Document.cpp:634]: Now implemented using explicit `to_lowecase` and `is_one_of` where applicable. Thanks for the suggestions!

File: Userland/Libraries/LibWeb/DOM/Document.cpp [hunk 4]
```diff
@@ -628,6 +630,61 @@ NonnullRefPtr<Range> Document::create_range()
...
+{
+    auto interface_lowecase = interface.to_lowercase();
```
Comment (Lubrsi) [inline Userland/Libraries/LibWeb/DOM/Document.cpp:636]: Typo:
```suggestion
    auto interface_lowercase = interface.to_lowercase();
```

File: Userland/Libraries/LibWeb/DOM/Document.cpp [hunk 5]
```diff
@@ -628,6 +630,61 @@ NonnullRefPtr<Range> Document::create_range()
...
+        event = Event::create(""); // FIXME: Create BeforeUnloadEvent
+    } else if (interface_lowecase == "compositionevent") {
+        event = Event::create(""); // FIXME: Create CompositionEvent
+    } else if (interface_lowecase == "customevent") {
+        event = Event::create(""); // FIXME: Create CustomEvent
+    } else if (interface_lowecase == "devicemotionevent") {
+        event = Event::create(""); // FIXME: Create DeviceMotionEvent
+    } else if (interface_lowecase == "deviceorientationevent") {
+        event = Event::create(""); // FIXME: Create DeviceOrientationEvent
+    } else if (interface.equals_ignoring_case("dragevent")) {
```
Comment (Lubrsi) [inline Userland/Libraries/LibWeb/DOM/Document.cpp:648]: Accidentally forgot to use the new `interface_lowercase`:
```suggestion
    } else if (interface_lowercase == "dragevent")) {
```

File: Userland/Libraries/LibWeb/DOM/Document.cpp [hunk 6]
```diff
@@ -628,6 +630,61 @@ NonnullRefPtr<Range> Document::create_range()
...
+        event = Event::create(""); // FIXME: Create DragEvent
+    } else if (interface.is_one_of("event", "events")) {
```
Comment (Lubrsi) [inline Userland/Libraries/LibWeb/DOM/Document.cpp:650]: Ditto:
```suggestion
    } else if (interface_lowercase.is_one_of("event", "events")) {
```

File: Userland/Libraries/LibWeb/DOM/Document.cpp [hunk 7]
```diff
@@ -628,6 +630,61 @@ NonnullRefPtr<Range> Document::create_range()
...
+        event = Event::create(""); // FIXME: Create FocusEvent
+    } else if (interface_lowecase == "hashchangeevent") {
+        event = Event::create(""); // FIXME: Create HashChangeEvent
+    } else if (interface_lowecase == "htmlevents") {
+        event = Event::create("");
+    } else if (interface_lowecase == "keyboardevent") {
+        event = Event::create(""); // FIXME: Create KeyboardEvent
+    } else if (interface_lowecase == "messageevent") {
+        event = Event::create(""); // FIXME: Create MessageEvent
+    } else if (interface.is_one_of("mouseevent", "mouseevents")) {
```
Comment (Lubrsi) [inline Userland/Libraries/LibWeb/DOM/Document.cpp:662]: Ditto:
```suggestion
    } else if (interface_lowercase.is_one_of("mouseevent", "mouseevents")) {
```

File: Userland/Libraries/LibWeb/DOM/Document.cpp [hunk 8]
```diff
@@ -628,6 +630,61 @@ NonnullRefPtr<Range> Document::create_range()
...
+        event = UIEvents::MouseEvent::create("", 0, 0);
+    } else if (interface_lowecase == "storageevent") {
+        event = Event::create(""); // FIXME: Create StorageEvent
+    } else if (interface_lowecase == "svgevents") {
+        event = Event::create("");
+    } else if (interface_lowecase == "textevent") {
+        event = Event::create(""); // FIXME: Create CompositionEvent
+    } else if (interface_lowecase == "touchevent") {
+        event = Event::create(""); // FIXME: Create TouchEvent
+    } else if (interface.is_one_of("uievent", "uievents")) {
```
Comment (Lubrsi) [inline Userland/Libraries/LibWeb/DOM/Document.cpp:672]: Ditto:
```suggestion
    } else if (interface_lowercase.is_one_of("uievent", "uievents")) {
````

**Comment by awesomekling** [comment]
This looks awesome. :)

**Pull request merged**